byron/ledger/executable-spec/test/Ledger/Upiec/Properties.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Ledger.Upiec.Properties
+  ( noProtVerChange
+  , protVerChangeAdopt
+  , protVerChangeSameComponents
+  , protVerChangeEmptyComponents
+  ) where
+
+import           Control.Lens (_1, _2, _3, _4, _5, _6, _7, _8, _9, (^.))
+import           Control.State.Transition (applySTS, TRC(..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import           Ledger.Core (Slot)
+import           Ledger.Update (UpId)
+import qualified Ledger.Update.Generators as G
+import           Hedgehog
+import           Ledger.Update (PVBUMP, UPIEC, ProtVer, PParams)
+
+
+-- Property #1 for the UPIEC STS
+--
+-- If PVBUMP transitions into a state with the same protocol version,
+-- the UPIEC STS results in the same state it started in.
+noProtVerChange :: Property
+noProtVerChange = property $ do
+  jc <- forAll G.noProtVerChangeJC
+  let (_, st, _) = jc
+  case applySTS @UPIEC (TRC jc) of
+    Right st' -> st === st'
+    Left _    -> failure
+
+-- Property #2 for the UPIEC STS
+--
+-- If PVBUMP transitions into a state with a new different protocol
+-- version, the UPIEC STS results in a state that adopts the new
+-- protocol version (pv) and new protocol parameters (pps).
+protVerChangeAdopt :: Property
+protVerChangeAdopt = property $ do
+  jc <- forAll G.protVerChangeJC
+  let
+    (s_n, st, _) = jc
+    (pv, pps)    = st ^. _1 :: (ProtVer, PParams)
+    fads         = st ^. _2 :: [(Slot, (ProtVer, PParams))]
+  case applySTS @PVBUMP (TRC ((s_n, fads), (pv, pps), ())) of
+    Left _            -> failure
+    Right (pv', pps') -> do
+      case applySTS @UPIEC (TRC jc) of
+        Right st' -> do
+          let (pvst', ppsst') = st' ^. _1 :: (ProtVer, PParams)
+          pv   /== pvst'
+          pv'  === pvst'
+          pps' === ppsst'
+        Left _    -> failure
+
+-- Property #3 for the UPIEC STS
```

Comment:
I don't see why do we need these other properties either. In property `#1` you stated something like: 

- If we generate a state which we know won't trigger a change in `PVBUMP`, when we shouldn't observe a change in `UPIEC`.

I would expect that we would only need a second property that states: "if we trigger a change in `PVBUMP`, then all these components in the final state change so and so ..."

Reply:
yes, I also vote for combining properties 2/3/4 into one in the executable spec

Reply:
@dnadales , are you saying that property #1 is not needed? What property would you have for a case when `pv = pv'`?

@uroboros, I think Nadales is talking about a different thing here, but your vote is noted!

Reply:
@uroboros, I think Nadales is talking about a different thing here, but your vote is noted!

The reason that I have them split into three different properties is that they talk about disjoint and unrelated parts of the resulting state: as far as `UPIEC` is considered, all the three different parts of the state have nothing to do with each other.

Reply:
> @dnadales , are you saying that property `#1` is not needed? What property would you have for a case when pv = pv'?

 No, I'm saying that I'm OK with having property `#1`. And if we have it, then we should have a similar property for the other case, and that's all we should need ...

Reply:
> The reason that I have them split into three different properties is that they talk about disjoint and unrelated parts of the resulting state: as far as UPIEC is considered, all the three different parts of the state have nothing to do with each other.

I'm not following this, sorry. Care to elaborate?

Reply:
Some parts of the state stay the same on `pv != pv'` (property no. 3 in the code) and some parts get reset (property no. 4). One part, in particular the `(pv, pps)` component, gets updated (property no. 2).